packages/core/src/modules/connections/DidExchangeProtocol.ts

```diff
@@ -0,0 +1,462 @@
+
+    const peerDid = DidPeer.fromDid(message.did)
+    let didDocument
+    if (peerDid.numAlgo === PeerDidNumAlgo.GenesisDoc) {
+      if (!message.didDoc) {
+        throw new AriesFrameworkError('No did doc')
+      }
+      // Verify signature on DidDoc attachment and assign DidDoc to connection record
+      await this.verifyAttachmentSignature(message.didDoc, connectionRecord.invitation?.recipientKeys)
+      didDocument = JsonTransformer.fromJSON(message.didDoc.getDataAsJson(), DidDocument)
+    } else {
+      didDocument = peerDid.didDocument
+    }
+
+    const didRecord = new DidRecord({
+      id: message.did,
+      role: DidDocumentRole.Received,
+      didDocument,
+      tags: {
+        // We need to save the recipientKeys, so we can find the associated did
+        // of a key when we receive a message from another connection.
+        recipientKeys: didDocument.recipientKeys,
+      },
+    })
+
+    await this.didRepository.save(didRecord)
+
+    connectionRecord.theirDid = message.did
+
+    await this.updateState(DidExchangeResponseMessage.type, connectionRecord)
+    this.logger.debug(`Process message ${DidExchangeResponseMessage.type} end`, connectionRecord)
+    return connectionRecord
+  }
+
+  public async createComplete(connectionRecord: ConnectionRecord): Promise<DidExchangeCompleteMessage> {
+    this.logger.debug(`Create message ${DidExchangeCompleteMessage.type} start`, connectionRecord)
+    DidExchangeStateMachine.assertCreateMessageState(DidExchangeCompleteMessage.type, connectionRecord)
+
+    const threadId = connectionRecord.threadId
+    const parentThreadId = connectionRecord.invitation?.id
+
+    if (!threadId) {
+      throw new AriesFrameworkError(`Connection record ${connectionRecord.id} does not have 'threadId' attribute.`)
+    }
+
+    if (!parentThreadId) {
+      throw new AriesFrameworkError(
+        `Connection record ${connectionRecord.id} does not have 'parentThreadId' attribute.`
+      )
+    }
+
+    const message = new DidExchangeCompleteMessage({ threadId, parentThreadId })
+
+    await this.updateState(DidExchangeCompleteMessage.type, connectionRecord)
+    this.logger.debug(`Create message ${DidExchangeCompleteMessage.type} end`, { connectionRecord, message })
+    return message
+  }
+
+  public async processComplete(
+    messageContext: InboundMessageContext<DidExchangeCompleteMessage>
+  ): Promise<ConnectionRecord> {
+    this.logger.debug(`Process message ${DidExchangeCompleteMessage.type} start`, messageContext)
+    const { connection: connectionRecord, message } = messageContext
+
+    if (!connectionRecord) {
+      throw new AriesFrameworkError('No connection record in message context.')
+    }
+
+    DidExchangeStateMachine.assertProcessMessageState(DidExchangeCompleteMessage.type, connectionRecord)
+
+    if (connectionRecord.invitation?.id !== message.thread?.parentThreadId) {
```

Comment:
Do we only need to check for parentThreadId?

Reply:
Good catch. I added check for `threadId` and also added the same check to `processResponse` method.